design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,434 @@
+# Progress reporting for backups and restores handled by volume snapshotters
+
+Users face difficulty in knowing the progress of backup/restore operations of volume snapshotters. This is very similar to the issues faced by users to know progress for restic backup/restore, like, estimation of operation, operation in-progress/hung etc.
+
+Each plugin might be providing a way to know the progress, but, it need not uniform across the plugins.
+
+Even though plugins provide the way to know the progress of backup operation, this information won't be available to user during restore time on the destination cluster.
+
+So, apart from the issues like progress, status of operation, volume snapshotters have unique problems like
+- not being uniform across plugins
+- not knowing the backup information during restore operation
+- need to be optional as few plugins may not have a way to provide the progress information
+
+This document proposes an approach for plugins to follow to provide backup/restore progress, which can be used by users to know the progress.
+
+## Goals
+
+- Provide uniform way of visibility into backup/restore operations performed by volume snapshotters
+
+## Non Goals
+- Plugin implementation for this approach
+
+## Background
+
+(Omitted, see introduction)
+
+## High-Level Design
+
+### Progress of backup operation handled by volume snapshotter
+
+Progress will be updated by volume snapshotter in VolumePluginBackup CR which is specific to that backup operation.
```

Comment:
I'm just going back and forth on the API design a bit. I think either option is workable for now for this specific use case, just trying to think through other uses/future extensions and whether that should push us in one direction or the other.

- Approach 1 combines the snapshot & backup info into a single CRD, and then has a parallel CRD on the restore side
- Approach 2 models the backup separately from the snapshot (the snapshot remains as an internal struct, not a CRD), but there's a lot of duplication of info between the snapshot & backup structs. On the restore side, there's one struct that models both the restore from durable storage, and the snapshot restore.
- we could consider fully breaking things out, so 4 CRDs: on the backup side, one for the snapshot, one for the backup to durable storage; on the restore side, one for the restore from durable storage, and one for the snapshot restore

I think I'm still leaning towards approach 1, as it solves the need and is simplest overall, but I'd like other maintainers' inputs on this (including any other options you see).

Reply:
I think I'm leaning towards approach 1, too, with the caveat that it needs to be optional. We can't require it if not all plugins are updating these CRs.

Reply:
thats true @nrb .. With approach 1, a CR will be created for every backup/restore operation even though plugin doesn't update it.